04.ls/ls.rb

```diff
@@ -1,18 +1,63 @@
 #!/usr/bin/env ruby
 
-dir = Dir.glob '*'
-FILES = dir.count
-COLUMN = 3
-ROW = FILES.ceildiv(COLUMN)
+require 'optparse'
+require 'etc'
+
+def chmod(mode)
+  mode.chars.map do |num|
+    %w[- --x -w- -wx r-- r-x rw- rwx][num.to_i]
+  end.join
+end
 
-MAX_ROW = dir.max_by(&:length).length + 2
+def list_detailed(dir)
+  dir.each do |name|
+    info = File.stat(name)
+    mode = info.mode.to_s(8)[-4..]
+    chmod = chmod(mode)
+    u_name = Etc.getpwuid(info.uid).name
+    group = Etc.getgrgid(info.gid).name
+    puts "#{chmod}  #{info.nlink} #{u_name}  #{group}  #{info.size}  #{info.mtime.mon} #{info.mtime.day} #{info.mtime.hour}:#{info.mtime.min} #{name}"
+  end
+end
 
-index = 0
-1.upto(ROW) do |row| # 行ごとに繰り返し
-  COLUMN.times do # 列ごとに繰り返し
-    print dir[index].to_s.ljust(MAX_ROW, ' ') unless dir[index].nil?
-    index += ROW # 行数の倍数 == 配列から取得するindex変数の値
+def list(dir, num_row, max_row)
+  index = 0
+  1.upto(num_row) do |row| # 行ごとに繰り返し
+    COLUMN.times do # 列ごとに繰り返し
+      print dir[index].to_s.ljust(max_row, ' ') unless dir[index].nil?
+      index += num_row # 配列から取得するindex変数の値 == 行数の倍数
+    end
+    index = row
+    print "\n"
   end
-  index = row
-  print "\n"
 end
+
+def get_dir(options)
+  second_arg = options[:a] == true ? File::FNM_DOTMATCH : 0
+  Dir.glob('*', second_arg)
+end
+
+def files_count(dir)
+  dir.count
+end
+
+def num_row(dir)
+  files_count(dir).ceildiv(COLUMN)
+end
+
+def max_row(dir)
+  dir.max_by(&:length).length + 2
+end
+
+COLUMN = 3
+
+options = {}
+OptionParser.new do |opt|
+  opt.on('-a') { |v| options[:a] = v }
+  opt.on('-r') { |v| options[:r] = v }
+  opt.on('-l') { |v| options[:l] = v }
+end.parse!
+
+dir = options[:r] == true ? get_dir(options).reverse : get_dir(options)
```

Comment:
こちら、lsコマンド3の時点では以下のように書かれていますね。

https://github.com/aki366/ruby-practices/pull/11/files

```ruby
dir = dir.reverse if option == '-r'
```

このように後置ifを使った方がスッキリしていて良さそうです。あと、 `== true` との比較は省略しても同じ結果になるので、そのようにしてください。修正をお願いいたします 🙏

Reply:
下記のように2行に分けてifを使うように修正してみました。
```suggestion
dir = get_dir(options)
dir = dir.reverse if options[:r]
```
`== true` 省略できるんですね。ご指摘ありがとうございます🙏
こちらも修正しましたので、確認お願いします 🙇